### PR TITLE
#19587 Dot Theme Selector Dropdown

### PR DESCRIPTION
**Pull request by ivorpad**

### CODE REVIEW

File: src/app/portlets/shared/dot-content-types-listing/dot-content-types-listing.module.ts [hunk 1]
```diff
@@ -8,14 +8,19 @@ import { DotListingDataTableModule } from '@components/dot-listing-data-table';
 import { DotBaseTypeSelectorModule } from '@components/dot-base-type-selector';
```
Comment (fmontes) [inline src/app/portlets/shared/dot-content-types-listing/dot-content-types-listing.module.ts:8]: Needs to clean this out.

File: src/app/view/components/_common/searchable-dropdown/component/searchable-dropdown.component.ts [hunk 2]
```diff
@@ -325,6 +337,10 @@ export class SearchableDropdownComponent
         }
     }
 
+    private isModifierKey(key: string) {
```
Comment (fmontes) [inline src/app/view/components/_common/searchable-dropdown/component/searchable-dropdown.component.ts:340]:
```suggestion
    private isModifierKey(key: string): boolean {
```

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            provide: NG_VALUE_ACCESSOR,
+            useExisting: forwardRef(() => DotThemeSelectorDropdownComponent)
+        }
+    ]
+})
+export class DotThemeSelectorDropdownComponent implements OnInit, ControlValueAccessor {
+    themes: DotTheme[] = [];
+    value: DotTheme = null;
+    totalRecords: number = 0;
+    placeholder = 'Select Themes';
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:26]: This should be an input that we can change from outside.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<dot-searchable-dropdown
+    [(ngModel)]="selectedTheme"
+    (change)="onChange($event)"
+    (filterChange)="handleFilterChange($event)"
+    [placeholder]="placeholder"
+    [data]="themes"
+    [totalRecords]="totalRecords"
+    (pageChange)="handlePageChange($event)"
+    rows="5"
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html:9]:
```suggestion
    rows="paginatorService.paginationPerPage"
```

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    labelPropertyName="name"
+    overlayWidth="350"
+    valuePropertyName="name"
+>
+    <!-- TODO: change overlayWidth to value + unit before it's merged -->
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html:14]: I think this change is already in master, if you merge master into this branch you should be good to do this change.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                    this.currentSiteIdentifier = identifier;
+                    this.paginatorService.setExtraParams('hostId', identifier);
+                    return this.paginatorService.getWithOffset(0).pipe(take(1));
+                })
+            )
+            .subscribe((themes) => {
+                this.themes = themes;
+                this.totalRecords = this.paginatorService.totalRecords;
+            });
+    }
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:56]:
```suggestion
        this.siteService
            .getCurrentSite()
            .pipe(
                pluck('identifier'),
                mergeMap((identifier) => {
                    this.currentSiteIdentifier = identifier;
                    this.paginatorService.setExtraParams('hostId', identifier);
                    return this.paginatorService.getWithOffset(0);
                })
            )
            .pipe(take(1))
            .subscribe((themes: DotTheme[]) => {
                this.themes = themes;
                this.totalRecords = this.paginatorService.totalRecords;
            });
    }
```

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        this.propagateChange = fn;
+    }
+
+    /**
+     * Writes a new value to the element
+     *
+     * @param {DotTheme} theme
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    writeValue(theme: DotTheme): void {
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:76]: We normally get from the endpoints (from example template endpoint) the theme id and for us to pass the `DotTheme` here we have to do an extra request outside this component.

This component should be smart enought to get the theme id and fetch the `DotTheme` itself.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    /**
+     * Handles the onChange behavior of the select input
+     *
+     * @param {DotTheme} { identifier }
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    onChange({ identifier }: DotTheme) {
+        this.themesService
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:87]: When this `onChange` is called, you already have the `DotTheme` you are doing an extra request to get it.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @param {DotTheme} theme
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    writeValue(theme: DotTheme): void {
+        this.value = theme;
+    }
+
+    /**
+     * Handles the onChange behavior of the select input
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:81]:
```suggestion
     * Handles the onChange from the searchable-dropdown
```

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    onChange({ identifier }: DotTheme) {
+        this.themesService
+            .get(identifier)
+            .pipe(take(1))
+            .subscribe((theme) => {
+                this.selectedTheme = theme.name;
+                this.writeValue(theme);
+            });
+        this.propagateChange(identifier);
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:94]: You are putting out an `identifier` but you expect a `DotTheme` in (see `writeValue()`)

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    handleFilterChange(filter: string): void {
+        this.getFilteredThemes(filter);
+    }
+
+    private getFilteredThemes(filter = '', offset = 0): void {
+        this.paginatorService.searchParam = filter;
+        this.paginatorService
+            .getWithOffset(this.currentOffset ? this.currentOffset : offset)
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:128]:
```suggestion
            .getWithOffset(this.currentOffset || offset)
```

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            .pipe(take(1))
+            .subscribe((themes) => {
+                this.themes = themes;
+                this.totalRecords = this.paginatorService.totalRecords;
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:132]: If you updates the `totalRecords` here, shouldn't you updated again in `handlePageChange`?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    handleFilterChange(filter: string): void {
+        this.getFilteredThemes(filter);
+    }
+
+    private getFilteredThemes(filter = '', offset = 0): void {
+        this.paginatorService.searchParam = filter;
+        this.paginatorService
+            .getWithOffset(this.currentOffset ? this.currentOffset : offset)
+            .pipe(take(1))
+            .subscribe((themes) => {
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:130]:
```suggestion
            .subscribe((themes: DotTheme[]) => {
```

**Review by fmontes — CHANGES_REQUESTED** [review]

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts (hunk 12 above)
Comment (ivorpad) [reply src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:132]: @fmontes No, `this.totalRecords` is set initially in `ngOnInit` and then in `getFilteredThemes` — right? so, when the results are filtered it runs again and sets `totalRecords` to the returned value from `this.paginatorService.totalRecords` and when you clear the input it runs again setting it to its initial value.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+        <ng-template #imageFallback>
+            <div class="dot-theme-item__image--fallback">
+                <span>{{ item.label.charAt(0) }}</span>
+            </div>
+        </ng-template>
+        <span class="dot-theme-item__meta">
+            <span class="dot-theme-item__label">{{ item.label }}</span>
+            <span class="dot-theme-item__date"
+                >Last modified: {{ item.modDate | date: 'MM/dd/yy' }}</span
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.html:22]: This needs message keys

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 15]
```diff
@@ -0,0 +1,55 @@
+  @use "variables"as *;
+  
+  .theme-selector__data-list-item {
+      cursor: pointer;
+      display: flex;
+      line-height: normal;
+      padding: 1rem 1rem;
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:7]: variable.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+      overflow: hidden;
+      text-overflow: ellipsis;
+      white-space: nowrap;
+      &:hover, &.selected {
+        background: $bg-highlight
+      }
+
+      .dot-theme-item__image,
+      .dot-theme-item__image--fallback {
+        border-radius: 4px;
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:19]: variable

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+      }
+      .dot-theme-item__image--fallback {
+        background: $bg-highlight;
+        width: 48px;
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:23]: can we use rem? for zoom and stuff?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+          line-height: 1;
+          vertical-align: middle;
+        }
+      }
+      .dot-theme-item__meta {
+            display: flex;
+            flex-direction: column;
+            justify-content: center;
+            margin-left: $spacing-2;
+          .dot-theme-item__label {
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:40]: This file looks like prettier doesn't run on it.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+            font-weight: bold;
+            display: inline-block;
+            margin-bottom: $spacing-1;
+          }
+          .dot-theme-item__date {
+            color: $gray-dark;
+            font-size: $font-size-small;
+          }
+          &>span {
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:49]:
```suggestion

          & > span {
```

Can you check is prettier is running on `scss` files?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+})
+export class DotThemeSelectorDropdownComponent implements OnInit, ControlValueAccessor {
+    themes: DotTheme[] = [];
+    value: DotTheme = null;
+    totalRecords: number = 0;
+    currentSiteIdentifier: string;
+    currentOffset: number;
+
+    @Input()
+    placeholder = 'Select Themes';
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:30]: Needs messages keys.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+    ) {}
+
+    ngOnInit(): void {
+        this.paginatorService.url = 'v1/themes';
+        this.paginatorService.paginationPerPage = 5;
+
+        this.siteService
+            .getCurrentSite()
+            .pipe(
+                take(1),
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:45]: I;m not sure if this `take(1)` is being applied to the `this.paginatorService.getWithOffset` as well.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+     * Writes a new value to the element
+     *
+     * @param {DotTheme} theme
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    writeValue(identifier: string): void {
+        this.themesService
+            .get(identifier)
+            .pipe(take(1))
+            .subscribe((theme) => {
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:81]: type.

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+     * @memberof DotThemeSelectorDropdownComponent
+     */
+    handleFilterChange(filter: string): void {
+        this.getFilteredThemes(filter);
+    }
+
+    private getFilteredThemes(filter = '', offset = 0): void {
+        this.paginatorService.searchParam = filter;
+        this.paginatorService
+            .getWithOffset(this.currentOffset ? this.currentOffset : offset)
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:128]:
```suggestion
            .getWithOffset(this.currentOffset || offset)
```

File: src/app/view/components/dot-theme-selector-dropdown/stories-data/data.json [hunk 24]
```diff
@@ -0,0 +1,802 @@
+[
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/stories-data/data.json:1]: What is this for?

File: src/app/view/pipes/dot-message/dot-message-pipe.module.ts [hunk 25]
```diff
@@ -1,10 +1,12 @@
...
 import { CommonModule } from '@angular/common';
 import { DotMessagePipe } from './dot-message.pipe';
+import { DotMessageService } from '@services/dot-message/dot-messages.service';
 
 @NgModule({
     imports: [CommonModule],
     declarations: [DotMessagePipe],
-    exports: [DotMessagePipe]
+    exports: [DotMessagePipe],
+    providers: [DotMessageService]
```
Comment (fmontes) [inline src/app/view/pipes/dot-message/dot-message-pipe.module.ts:10]: Why?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+    ngOnInit(): void {
+        this.paginatorService.url = 'v1/themes';
+        this.paginatorService.paginationPerPage = 5;
+
+        this.siteService
+            .getCurrentSite()
+            .pipe(
+                take(1),
+                pluck('identifier'),
+                mergeMap((identifier) => {
```
Comment (alfredo-dotcms) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:47]: type?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+                    this.currentSiteIdentifier = identifier;
+                    this.paginatorService.setExtraParams('hostId', identifier);
+                    return this.paginatorService.getWithOffset(0).pipe(take(1));
+                })
+            )
+            .subscribe((themes) => {
```
Comment (alfredo-dotcms) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:53]: type?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+     * @memberof SearchableDropdownComponent
+     */
+    registerOnChange(fn): void {
+        this.propagateChange = fn;
+    }
+
+    /**
+     * Writes a new value to the element
+     *
+     * @param {DotTheme} theme
```
Comment (alfredo-dotcms) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:74]: update doc

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    onChange(theme: DotTheme) {
+        this.value = theme;
+        this.propagateChange(theme.identifier);
+    }
+    /**
+     * Fetches a list of themes using an offset for pagination
+     *
+     * @param {LazyLoadEvent} event
+     * @return void
```
Comment (alfredo-dotcms) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:100]: not needed

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    .dot-theme-item__image--fallback {
+        background: $bg-highlight;
+        width: 3.43rem;
+        height: 3.43rem;
+        font-size: $font-size-xx-large;
+        color: $brand-primary_mod;
+        text-transform: uppercase;
+        text-align: center;
+        line-height: 45px;
```
Comment (fmontes) [inline src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.scss:30]: rems?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts (hunk 26 above)
Comment (fmontes) [reply src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:46]: ?

File: src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts (hunk 9 above)
Comment (fmontes) [reply src/app/view/components/dot-theme-selector-dropdown/dot-theme-selector-dropdown.component.ts:86]: ?

**Review by fmontes — CHANGES_REQUESTED** [review]

**Pull request merged**